server/src/main/java/org/opensearch/index/translog/transfer/TranslogTransferManager.java

```diff
@@ -218,9 +216,9 @@ public void deleteGenerationAsync(long primaryTerm, Set<Long> generations) {
             metadataFiles.add(metadataFileName);
         });
         // Delete the translog and checkpoint files asynchronously
-        deleteTranslogFilesAsync(primaryTerm, translogFiles);
+        deleteTranslogFilesAsync(primaryTerm, translogFiles, onDeletion);
```

Comment:
should we have two semaphores in `TransferManager` instead - one for files and one for metadata ?

Reply:
Having single semaphore ensures that generationsToDelete is consistent and so is deletion of translog & metadata files. If we had to 2, it is possible that translog deletion has not completed while metadata deletion has completed. Since translog deletion has not completed, the translogTransferTracker still has the generation and this will lead a request which had metadata files to be deleted that have already been deleted.
https://github.com/opensearch-project/OpenSearch/blob/87bdd0806b77ea4c0ae1eed6830fd1570abd36f6/server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java#L346-L352

Reply:
There is no need to consistency here , since delete is idempotent .  Why I am suggesting is that it would move the management responsibility to entirely `TransferManager` .